src/net_processing.cpp

```diff
@@ -3036,6 +3045,112 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     return;
 }
 
+PeerManagerImpl::InvalidTxTask PeerManagerImpl::ProcessInvalidTx(const CTransactionRef& tx, NodeId nodeid, const TxValidationState& state)
+{
+    AssertLockHeld(cs_main);
+    AssertLockNotHeld(m_peer_mutex);
+    LogPrint(BCLog::MEMPOOLREJ, "%s (wtxid=%s) from peer=%d was not accepted: %s\n",
+             tx->GetHash().ToString(),
+             tx->GetWitnessHash().ToString(),
+             nodeid,
+             state.ToString());
+    // Maybe punish peer that gave us an tx
+    MaybePunishNodeForTx(nodeid, state);
+
+    switch (state.GetResult()) {
+    case TxValidationResult::TX_RESULT_UNSET:
+    case TxValidationResult::TX_NO_MEMPOOL:
+    {
+        // This function should only be called when a transaction fails validation.
+        Assume(false);
+        return InvalidTxTask::NONE;
+    }
+    case TxValidationResult::TX_UNKNOWN:
+    {
+        // Transaction was not validated; we don't know that it is invalid. Do not add it to any
+        // rejection caches or forget about it yet.
+        return InvalidTxTask::NONE;
+    }
+    case TxValidationResult::TX_WITNESS_STRIPPED:
+    {
+        // Do not add txids of witness transactions or witness-stripped
+        // transactions to the filter, as they can have been malleated;
+        // adding such txids to the reject filter would potentially
+        // interfere with relay of valid transactions from peers that
+        // do not support wtxid-based relay. See
+        // https://github.com/bitcoin/bitcoin/issues/8279 for details.
```

Comment:
@sdaftuar honestly this comment/link doesn't make sense to me. If it's witness-stripped, we can't add it to `m_recent_rejects` since that would blind us to any real version of the tx?

Reply:
I had to remind myself of how this works, but the concern contemplated in that comment is that if the **only** copy of a non-witness-stripped version of a transaction is from a node that has not yet upgraded to wtxid-based-relay, then an adversary could blind us to that transaction by relaying a witness-stripped version of it (and then, since txid==wtxid for witness-stripped transactions, the hash that we'd use to fetch the transaction from the non-wtxid-relay peer who has the correct version would be in our `m_recent_rejects`, and we wouldn't ever request it).

At the time that I wrote that comment, I think I also said that once wtxid-based-relay is sufficiently deployed, that we could stop worrying about this...  I don't recall when wtxid-relay was deployed but maybe we're at that point already?

There's another link to a comment thread in the PR that implemented wtxid-relay which discusses the potential impact on downstream projects, so maybe if we were to make a behavior change we should communicate that in advance so that no one is surprised.

Reply:
It seems like the comment thread is concerned with what happens if:
(1) the sender is using non-wtxidrelay (and thus only announces transactions by txid),
(2) all of its peers are caching witness-stripped rejections by wtxid (i.e. txid),
(3) there is an attacker trying to censor the sender's tx by sending witness-stripped versions of the tx ahead of the sender.

This isn't a problem if (1) isn't true, i.e. the sender is using wtxidrelay, e.g. any non-eol version of Bitcoin Core. If we see that the vast majority of nodes (that send transactions) are sending a `wtxidrelay` message, perhaps it would be safe to remove the special casing (with a message to downstream projects)?

Reply:
Resolving as out of scope for this PR. Also fwiw, btcd doing announcements by txid seems to be a reason to keep it for now...